openssl/rsa.go

```diff
@@ -170,18 +170,47 @@ func EncryptRSANoPadding(pub *PublicKeyRSA, msg []byte) ([]byte, error) {
 	return evpEncrypt(pub.withKey, C.GO_RSA_NO_PADDING, nil, nil, msg)
 }
 
+func saltLength(golen int, sign bool) (C.int, error) {
+	// A salt length of -2 is valid in OpenSSL, but not in crypto/rsa, so reject
+	// it, and lengths < -2, before we convert to the OpenSSL sentinel values.
+	if golen <= -2 {
+		return 0, errors.New("crypto/rsa: PSSOptions.SaltLength cannot be negative")
+	}
+	clen := C.int(golen)
+	// OpenSSL uses sentinel salt length values like Go do,
+	// but the values don't fully match for rsa.PSSSaltLengthAuto (0).
+	if golen == 0 {
+		if sign {
+			if vMajor == 1 {
+				// OpenSSL 1.x uses -2 to mean maximal size when signing where Go use 0.
+				clen = C.GO_RSA_PSS_SALTLEN_MAX_SIGN
+			} else {
+				// OpenSSL 3.x deprecated RSA_PSS_SALTLEN_MAX_SIGN
+				// and uses -3 to mean maximal size when signing where Go use 0.
+				clen = C.GO_RSA_PSS_SALTLEN_MAX
+			}
+		} else {
+			// OpenSSL uses -2 to mean auto-detect size when verifying where Go use 0.
+			clen = C.GO_RSA_PSS_SALTLEN_AUTO
+		}
+	}
+	return clen, nil
+}
+
 func SignRSAPSS(priv *PrivateKeyRSA, h crypto.Hash, hashed []byte, saltLen int) ([]byte, error) {
-	if saltLen == 0 {
-		saltLen = -1 // RSA_PSS_SALTLEN_DIGEST
+	slen, err := saltLength(saltLen, true)
```

Comment:
This seems like a lot of things that only vary by how many letters of `saltLength` they chop out. 😄 This is what comes to mind:

```suggestion
	opensslSaltLen, err := convertGoSaltLength(saltLen, true)
```

But... these functions are short, and the types are different, so there's not much chance to actually mix them up and I think it's fine as-is if you prefer.

Reply:
Too mach chopping 🥢. I've simplified `saltLength` and changed a couple of names to make it clearer.